src/lib/wallet/GoodWalletProvider.js

```diff
@@ -34,36 +35,89 @@ export const GoodWalletProvider = ({ children, disableLoginAndWatch = false }) =
   const [dailyUBI, setDailyUBI] = useState()
   const [isCitizen, setIsCitizen] = useState()
   const [shouldLoginAndWatch] = usePropsRefs([disableLoginAndWatch === false])
-  const lastLoginRef = useRef(null)
+
   const db = getDB()
 
-  // when new wallet set the web3provider for future use with usedapp
-  useEffect(() => {
-    if (!goodWallet) {
-      return
+  const verifyUserIsRegistered = async walletLogin => {
+    const { decoded, jwt } = await walletLogin.validateJWTExistenceAndExpiration()
+    const { aud } = decoded || {}
+
+    log.info('verifyUserIsRegistered: jwt data', { decoded, jwt, aud })
+
+    if (!decoded || aud === 'unsigned') {
+      throw new Error('jwt is of unsigned user', 'UnsignedJWTError')
     }
+  }
+  const update = useCallback(
+    async goodWallet => {
+      const calls = [
+        {
+          balance: goodWallet.tokenContract.methods.balanceOf(goodWallet.account),
+        },
+        {
+          ubi: goodWallet.UBIContract.methods.checkEntitlement(goodWallet.account),
+        },
+        {
+          isCitizen: goodWallet.identityContract.methods.isWhitelisted(goodWallet.account),
+        },
+      ]
+
+      // entitelment is separate because it depends on msg.sender
+      const [[{ balance }, { ubi }, { isCitizen }]] = await goodWallet.multicallFuse.all([calls])
 
-    setWeb3(new HDWalletProvider(goodWallet.accounts, goodWallet.wallet._provider.host))
-  }, [goodWallet, setWeb3])
+      setBalance(parseInt(balance))
+      setDailyUBI(parseInt(ubi))
+      setIsCitizen(isCitizen)
+    },
+    [setBalance, setDailyUBI, setIsCitizen],
+  )
 
   const initWalletAndStorage = useCallback(
-    async (seedOrWeb3, type: 'SEED' | 'METAMASK' | 'WALLETCONNECT' | 'OTHER', initRegistered = false) => {
+    async (seedOrWeb3, type: 'SEED' | 'WEB3WALLET') => {
```

Comment:
needs also to be merged with my suggestions from https://github.com/GoodDollar/GoodDAPP/pull/3808

Reply:
done